# Handling Metadata with MuseScore and Music21

First we need to import the music21 toolkit.

In [1]:
import music21

Next, we parse one of our in MusicXML encoded files.

In [10]:
chant = music21.converter.parse('Corpus/Monophonic/Book-2_Emendemus-in-melius_p-142.xml')

In [11]:
# @TODO: fix problem with MusicXML encoding here ...
chant.show('text')

{0.0} <music21.text.TextBox "Anonymous">
{0.0} <music21.text.TextBox "Emendemus ...">
{0.0} <music21.metadata.Metadata object at 0x108c04c50>
{0.0} <music21.stream.Part Piano>
    {0.0} <music21.instrument.Instrument P1: Piano: Piano>
    {0.0} <music21.stream.Measure 1 offset=0.0>
        {0.0} <music21.layout.SystemLayout>
        {0.0} <music21.clef.BassClef>
        {0.0} <music21.key.KeySignature of no sharps or flats, mode none>
        {0.0} <music21.meter.TimeSignature 63/4>
        {0.0} <music21.note.Note A>
        {1.0} <music21.note.Note A>
        {2.0} <music21.note.Note C>
        {3.0} <music21.note.Note D>
        {4.0} <music21.note.Note F>
        {5.0} <music21.note.Note D>
        {6.0} <music21.note.Note F>
        {7.0} <music21.note.Note F>
        {8.0} <music21.note.Note E>
        {9.0} <music21.note.Note F>
        {10.0} <music21.note.Note D>
        {11.0} <music21.note.Note E>
        {12.0} <music21.note.Note D>
        {13.0} <music21.note.Note C>
    

After we parsed the MusicXML file into an music21 stream, we want to extract what type of meta data is available to us, 
by creating a list.

In [12]:
chant_md = chant.metadata
chant_md_list = chant_md.all()
chant_md_list

[('None', 'Hypodorian'),
 ('arranger', 'Glarean'),
 ('composer', 'Anonymous'),
 ('movementName', 'Emendemus in melius'),
 ('movementNumber', 'p. 103: p. 142'),
 ('number', 'Dodecachordon, Book 2'),
 ('title', 'Emendemus in melius')]

Accessing particular items in the by music21 generated list of tuplets is cumbersome. For example, let's access the composer:

In [13]:
chant_md_list[2][1]

'Anonymous'

However, if we convert our list of tuplets to a dictionary, we can look up the composer much more intuitively:

In [14]:
chant_md_dict = dict(chant_md_list)
chant_md_dict

{'None': 'Hypodorian',
 'arranger': 'Glarean',
 'composer': 'Anonymous',
 'movementName': 'Emendemus in melius',
 'movementNumber': 'p. 103: p. 142',
 'number': 'Dodecachordon, Book 2',
 'title': 'Emendemus in melius'}

In [15]:
chant_md_dict['composer']

'Anonymous'

In MuseScore the ```poet``` field under ```File > Score Properties``` renders to a ```None``` key. We use this key to
indicate the mode of the composition according to Glarean. Accessing the mode then only requires:

In [16]:
chant_md_dict['None']

'Hypodorian'

Here is an explanation of how all the fields from MuseScore's ```Score Properties``` are being used in the project:

MuseScore Label | Music21 Label | Value | Explanation
--------------- | ------------- | ----- | -----------
poet | 'None' | 'Hypodorian' | There is no default "Mode" field, so we can use this field instead.
arranger | 'arranger' | 'Glarean' | Glarean placed the composition as an example into his book.
composer | 'composer' | 'Anonymous' | Whoever composed the composition.
movementTitle | 'movementName' | 'Emendemus in melius' | Here the movement name is the same name as the 'title', since there are no additional movements.
movementNumber | 'movementNumber' | 'p. 103: p. 142' | The fields shows on what page of the _Dodecachordon_ the example appears. There are two numbers seperated by a colon: (1) the number before the colon indicates where the piece appears in the facsimile; (2) the number following the colon indicates where the piece appears in Miller's translation (see below for code snippet to seperate the two values).  
workNumber | 'number' | 'Dodecachordon, Book 2' | Indicates in which book of the dodecachordon the piece occurs.
workTitle | 'title' | 'Emendemus in melius' | The title of the piece.

In [17]:
# Code snippet to seperate the two page values:
chant_md_pages = chant_md_dict['movementNumber'].split(':')
chant_md_pages[0].strip() # Facsimile, and removing whitespace
chant_md_pages[1].strip() # Miller's translation, and removing whitespace

'p. 142'

In [18]:
chant2 = music21.converter.parse('Corpus/Monophonic/Book-2_Pater-noster_pp-143-44.xml')

In [19]:
chant2.metadata.all()

[('None', 'Aeolian'),
 ('arranger', 'Glarean'),
 ('composer', 'Anonymous'),
 ('movementName', 'Pater noster'),
 ('movementNumber', 'p. 105: p. 143-144'),
 ('number', 'Dodecachordon, Book 2'),
 ('title', 'Pater noster')]